testing/helpers/test-helpers.bash

```diff
@@ -30,3 +34,101 @@ function ensure_variables_set() {
         exit 1
     fi
 }
+
+# Converts the value of $COUCHBASE_SERVER_HOSTS into a file ready to feed to Prometheus file_sd.
+# Stores the name of the generated file in $PROMETHEUS_TARGETS_FILE.
+function _create_prometheus_targets_file() {
+    local tmp
+    tmp=$(mktemp -d)
+    echo "$COUCHBASE_SERVER_HOSTS" | jq -R -s '[split("\n")[:-1][] | sub(":8091"; ":9091")] | [{"targets": ., "labels": {"job":"couchbase-server","cluster":"smoke"}}]' > "$tmp/smoke.json"
+    export PROMETHEUS_TARGETS_FILE="$tmp/smoke.json"
+}
+
+# Starts a docker container named `cmos`, mounting $PROMETHEUS_TARGETS_FILE.
+# Exposes a variable $CMOS_HOST with the nginx host:port.
+# All parameters will be passed on to docker before the image.
+function _start_cmos() {
+    docker run --rm -d -p '8080' --name cmos -v "$PROMETHEUS_TARGETS_FILE:/etc/prometheus/couchbase/custom/smoke.json" "$@" "$CMOS_IMAGE"
+    local cmos_port
+    cmos_port=$(docker inspect cmos -f '{{with index .NetworkSettings.Ports "8080/tcp"}}{{ with index . 0 }}{{ .HostPort }}{{end}}{{end}}')
+    export CMOS_HOST="localhost:$cmos_port"
+}
+
+# Starts a Couchbase cluster and CMOS container.
+#
+# Parameters:
+# $SMOKE_NODES: The nubmer of nodes to start (defaults to 3)
+#
+# This function will set the following variables with its results:
+# $COUCHBASE_SERVER_HOSTS: the hostname/IP and management port of every CBS node, separated by newlines
+#   (Note that they may not be accessible from localhost, e.g. if running in a container - they'll be accessible to CMOS though)
+# $CMOS_HOST: the hostname/IP and nginx port of the running CMOS container
+#
+# Note: do not call this function using BATS run! Otherwise its variables will not be set.
+function start_smoke_cluster() {
+    local nodes=${SMOKE_NODES:-3}
+    echo "# Starting smoke cluster for platform $TEST_PLATFORM with $nodes nodes"
+    case $TEST_PLATFORM in
+        native)
+            export VAGRANT_NODES=$nodes
+            # TODO: move this into the matrix
+            start_vagrant_cluster "$COUCHBASE_SERVER_VERSION" "centos7"
+            wait_for_url 10 "$(echo "$COUCHBASE_SERVER_HOSTS" | head -n1)/ui"
+            _create_prometheus_targets_file
+            _start_cmos
+            ;;
+        containers)
+            ensure_variables_set CMOS_IMAGE
+            ensure_variables_set COUCHBASE_SERVER_IMAGE
+            docker network create cmos_test
```

Comment:
I'mm not sure I like this over a compose file, what's the reasoning?

Reply:
Compose file makes it difficult to have a variable number of nodes - we'd need to bring in templating or such